src/MarginModels/PointSetMarginModel.py

```diff
@@ -0,0 +1,124 @@
+    inputSpecs.addSub(InputData.parameterInputFactory('failedDataFileID', contentType=InputTypes.StringType, descr='failed data file'))
+
+    inputSpecs.addSub(InputData.parameterInputFactory('marginID', contentType=InputTypes.StringType, descr='ID of the margin variable'))
+
+    mapping = InputData.parameterInputFactory('map', contentType=InputTypes.StringType, descr='ID of the column of the csv containing failed data')
+    mapping.addParam("var", InputTypes.StringType)
+    inputSpecs.addSub(mapping)
+
+    return inputSpecs
+
+
+  def __init__(self):
+    """
+      Constructor
+      @ In, None
+      @ Out, None
+    """
+    MarginBase.__init__(self)
+
+    self.failedDataFileID = None  # name of the file containing the failed data
+    self.InvMapping = {}          # dictionary containing mapping between failed and actual data
+    self.marginID = None          # ID of the calculated margin variable
+    self.dimensionality = None    # dimensionality of the point set
+
+
+  def _handleInput(self, paramInput):
+    """
+      Function to read the portion of the parsed xml input that belongs to this specialized class
+      and initialize some stuff based on the inputs got
+      @ In, paramInput, InputData.ParameterInput, the parsed xml input
+      @ Out, None
+    """
+    super()._handleInput(paramInput)
+    for child in paramInput.subparts:
+      if child.getName() == 'failedDataFileID':
+        self.failedDataFileID = child.value
+      if child.getName() == 'marginID':
+        self.marginID = child.value
+      elif child.getName() == 'map':
+        self.InvMapping[child.value] = child.parameterValues.get('var')
+
+    self.dimensionality = len(self.InvMapping.values())
+
+
+  def initialize(self, inputDict):
+    """
+      Method to initialize this class
+      @ In, inputDict, dict, dictionary of inputs
+      @ Out, None
+    """
+    super().initialize(inputDict)
+    self.failedDataFileID = os.path.join(self.workingDir, self.failedDataFileID)
+    self.failedData = pd.read_csv(self.failedDataFileID)[self.InvMapping.values()]
+
+
+  def _marginFunction(self, inputDict):
+    """
+      Method to calculate margin value
+      @ In, inputDict, dict, dictionary of inputs
+      @ Out, outputDict, dict, dictionary containing margin value
+    """
+    actualData = pd.DataFrame(inputDict)
+    actualData = actualData.rename(columns=self.InvMapping)
+
+    distMatrix = pairwise_distances(self.failedData.values, actualData.values, metric=customDist)
+    distMatrix[distMatrix<0] = 0
+    margin = np.mean(distMatrix)
+
+    zeroPoint = copy.deepcopy(actualData)
+    zeroPoint[:] = 0.0
+    distMatrix2 = pairwise_distances(self.failedData.values, zeroPoint.values, metric=customDist)
+    distMatrix2[distMatrix2<0] = 0
+    margin2 = np.mean(distMatrix2)
+
+    outputDict = {}
+    outputDict[self.marginID] = margin/margin2
+
+    return outputDict
+
+
+def customDist(pointSet,refPoint):
```

Comment:
I think this function can be moved to utils in the future.

Reply:
Definitely, especially when we will push ER data analysis methods in math and nlp utils